README.md
# What is does

libapache2-mod-fastcgi has the annoying property to fail when AWS ELB
CNAME returns multiple ip addresses instead of simply picking the
first one.

The error message is something like:

  FastCgiExternalServer /usr/lib/cgi-bin/php5-fcgi: failed to resolve "internal-php-farm.ap-southeast-2.elb.amazonaws.com" to exactly one IP address

This improved version picks the first available ip address and sets a TTL
of 60 seconds. Every 60 seconds the DNS is queried for a new ip address.


# How to build

1. Make sure multiverse is enabled in /etc/apt/sources.list.
2. Make sure you have deb-src lines in /etc/apt/sources.list.
3. Install build tools
    sudo apt-get install devscripts
4. Install package build dependencies:
    sudo apt-get build-dep libapache2-mod-fastcgi
5. Make sure gcc 4.9 is installed and the default.
6. Build:
    dpkg-buildpackage -rfakeroot -b


debian/changelog
libapache-mod-fastcgi (2.4.7~0910052141-1.3local1) UNRELEASED; urgency=medium

  * Resolve hostname every 60 seconds if hostname instead of ip address
    is given (needs gcc 4.9 to compile).

 -- Berend de Boer <berend@xplainhosting.com>  Fri, 01 Jan 2016 13:05:01 +1300

libapache-mod-fastcgi (2.4.7~0910052141-1.1local1) UNRELEASED; urgency=medium

  * Xplain Hosting libapache2-mod-fastcgi

 -- Berend de Boer <berend@xplainhosting.com>  Tue, 07 Jul 2015 15:57:24 +1200

libapache-mod-fastcgi (2.4.7~0910052141-1.1) unstable; urgency=low

  * Non-maintainer upload.
  * Port to Apache 2.4 (closes: #666839).
  * Drop some more vestiges of Apache 1 support.

 -- Colin Watson <cjwatson@debian.org>  Thu, 11 Jul 2013 14:10:25 +0100

libapache-mod-fastcgi (2.4.7~0910052141-1) unstable; urgency=low

  * New upstream release

 -- Tatsuki Sugiura <sugi@nemui.org>  Mon, 15 Nov 2010 16:31:58 +0900

libapache-mod-fastcgi (2.4.6-1) unstable; urgency=low

  * New upstream release

 -- Tatsuki Sugiura <sugi@nemui.org>  Tue, 18 Mar 2008 11:23:42 +0900

libapache-mod-fastcgi (2.4.2-10) unstable; urgency=low

  * Drop apache1.3 module. (Closes: #429095)
  * Now this cannot be built with apache2.0.

 -- Tatsuki Sugiura <sugi@nemui.org>  Mon, 28 Jan 2008 15:17:58 +0900

libapache-mod-fastcgi (2.4.2-9) unstable; urgency=low

  * apply authn provider patch for apache2.2.
    Thanks for Martin Furter (Closes: #414185)
  * stop build apache 1.3 binary (Closes: #429095)
  * fix suexec wrapper path (Closes: #415913)

 -- Tatsuki Sugiura <sugi@nemui.org>  Sat, 25 Aug 2007 17:13:50 +0900

libapache-mod-fastcgi (2.4.2-8) unstable; urgency=high

  * transit libapache2-mod-fastcgi for apache2.2
  * Fix in built-in suexec path (Closes: #331617)
  * remove unnecessary libc6 version dependency (Closes: #343519)
  * apply patch to fix that apache2 segv on load (Closes: #343514)

 -- Tatsuki Sugiura <sugi@nemui.org>  Fri,  6 Oct 2006 20:53:21 +0900

libapache-mod-fastcgi (2.4.2-7) unstable; urgency=low

  * Update description. Thanks for Russ Allbery. (Closes: #320246)
  * a2{en,dis}cmd automatically called in postinst/prerm. (Closes: #320246)

 -- Tatsuki Sugiura <sugi@nemui.org>  Sat,  6 Aug 2005 14:18:30 +0900

libapache-mod-fastcgi (2.4.2-6) unstable; urgency=low

  * fix typo in original document (Closes: #288270)
  * add config topic in README.Debian (Closes: #298819)

 -- Tatsuki Sugiura <sugi@nemui.org>  Wed, 16 Mar 2005 22:27:02 +0900

libapache-mod-fastcgi (2.4.2-5) unstable; urgency=high

  * SuEXEC setting has been disabled in defualt. (Closes: #275201)

 -- Tatsuki Sugiura <sugi@nemui.org>  Tue, 19 Oct 2004 13:38:17 +0900

libapache-mod-fastcgi (2.4.2-4) unstable; urgency=low

  * Add Documentation

 -- Tatsuki Sugiura <sugi@nemui.org>  Wed,  8 Sep 2004 18:15:05 +0900

libapache-mod-fastcgi (2.4.2-3) unstable; urgency=high

  * Back out the ill-fated apache2 LFS transition.
  * Bump the apache2-{threaded,prefork}-dev build-dep to (>= 2.0.50-10)

 -- Tatsuki Sugiura <sugi@nemui.org>  Tue, 24 Aug 2004 22:44:11 +0900

libapache-mod-fastcgi (2.4.2-2) unstable; urgency=medium

  * rebuild for new ABI of apache2 2.0.50-9

 -- Tatsuki Sugiura <sugi@nemui.org>  Tue, 17 Aug 2004 15:08:04 +0900

libapache-mod-fastcgi (2.4.2-1) unstable; urgency=low

  * Upstream Release.
  * New Maintainer (Closes: #229214)

 -- Tatsuki Sugiura <sugi@nemui.org>  Sat, 24 Jul 2004 06:07:47 +0900

libapache-mod-fastcgi (2.4.0-4) unstable; urgency=low

  * QA upload.
  * libdb4.1-dev build-dependency removed. (Closes: #243017)
  * Standards-Version updated.

 -- Emanuele Rocca <ema@debian.org>  Tue, 13 Apr 2004 18:09:33 +0200

libapache-mod-fastcgi (2.4.0-3) unstable; urgency=low

  * Orphaned.

 -- Davide Puricelli (evo) <evo@debian.org>  Fri, 23 Jan 2004 19:15:54 +0100

libapache-mod-fastcgi (2.4.0-2) unstable; urgency=low

  * Removed a useless and buggy debian/postrm; closes: #181418.

 -- Davide Puricelli (evo) <evo@debian.org>  Wed, 19 Feb 2003 14:34:24 +0100

libapache-mod-fastcgi (2.4.0-1) unstable; urgency=low

  * New upstream version.
  * Bumping Standards-Version to 3.5.8.
  * Enhancing debian/postinst and debian/prerm scripts.

 -- Davide Puricelli (evo) <evo@debian.org>  Thu, 13 Feb 2003 14:21:57 +0100

libapache-mod-fastcgi (2.2.12-4) unstable; urgency=low

  * Fixed debian/rules (gah).

 -- Davide Puricelli (evo) <evo@debian.org>  Fri, 16 Aug 2002 14:45:48 +0200

libapache-mod-fastcgi (2.2.12-3) unstable; urgency=low

  * No longer links against (nonexistent) libdb.so.2.

 -- Davide Puricelli (evo) <evo@debian.org>  Fri, 16 Aug 2002 13:51:36 +0200

libapache-mod-fastcgi (2.2.12-2) unstable; urgency=low

  * Fixed -autoUpdate option, closes: #132981.
    Patch provided by Freddy Spierenburg <freddy@telegraafnet.nl>

 -- Davide Puricelli (evo) <evo@debian.org>  Sun, 17 Mar 2002 15:32:08 +0100

libapache-mod-fastcgi (2.2.12-1) unstable; urgency=low

  * New upstream version.

 -- Davide Puricelli (evo) <evo@debian.org>  Sun,  9 Dec 2001 21:07:42 +0000

libapache-mod-fastcgi (2.2.10-3) unstable; urgency=low

  * Now depends on apache-common 1.3.14-2.

 -- Davide Puricelli (evo) <evo@debian.org>  Thu,  7 Dec 2000 19:47:14 +0100

libapache-mod-fastcgi (2.2.10-2) unstable; urgency=low

  * Recompiled against apache-dev 1.3.14-1; closes: #78516, #78629.

 -- Davide Puricelli (evo) <evo@debian.org>  Sun,  3 Dec 2000 17:44:21 +0100

libapache-mod-fastcgi (2.2.10-1) unstable; urgency=low

  * New upstream version.

 -- Davide Puricelli (evo) <evo@debian.org>  Tue, 28 Nov 2000 19:40:33 +0100

libapache-mod-fastcgi (2.2.8-1) unstable; urgency=low

  * New upstream version.

 -- Davide Puricelli (evo) <evo@debian.org>  Mon, 28 Aug 2000 17:46:13 +0200

libapache-mod-fastcgi (2.2.6-2) unstable; urgency=low

  * Further fixed debian/postinst.

 -- Davide Puricelli (evo) <evo@debian.org>  Tue, 15 Aug 2000 12:06:59 +0200

libapache-mod-fastcgi (2.2.6-1) unstable; urgency=low

  * New upstream version.
  * New maintainer email address.


fcgi_util.c
/*
 * $Id: fcgi_util.c,v 1.32 2007/09/23 16:33:29 robs Exp $
 */

#include "fcgi.h"

#ifdef WIN32
#pragma warning( disable : 4100 )
#elif defined(APACHE2)
#include <netdb.h>
#include <unistd.h>
#include <grp.h>
#include <pwd.h>

#if APR_HAVE_ARPA_INET_H
#include <arpa/inet.h>
#endif

#include "unixd.h"
#endif

uid_t
fcgi_util_get_server_uid(const server_rec * const s)
{
#if defined(WIN32)
    return (uid_t) 0;
#elif defined(APACHE2)
    /* the main server's uid */
    return ap_user_id;
#else
    /* the vhost's uid */
    return s->server_uid;
#endif
}

uid_t
fcgi_util_get_server_gid(const server_rec * const s)
{
#if defined(WIN32)
    return (uid_t) 0;
#elif defined(APACHE2)
    /* the main server's gid */
    return ap_group_id;
#else
    /* the vhost's gid */
    return s->server_gid;
#endif
}

/*******************************************************************************
 * Compute printable MD5 hash. Pool p is used for scratch as well as for
 * allocating the hash - use temp storage, and dup it if you need to keep it.
 */
char *
fcgi_util_socket_hash_filename(pool *p, const char *path,
        const char *user, const char *group)
{
    char *buf = ap_pstrcat(p, path, user, group, NULL);

    /* Canonicalize the path (remove "//", ".", "..") */
    ap_getparents(buf);

    return ap_md5(p, (unsigned char *)buf);
}


 /*******************************************************************************
  * Concat src1 and src2 using the approprate path seperator for the platform.
  */
static char * make_full_path(pool *a, const char *src1, const char *src2)
{
#ifdef WIN32
    register int x;
    char * p ;
    char * q ;

    x = strlen(src1);

    if (x == 0) {
	    p = ap_pstrcat(a, "\\", src2, NULL);
    }
    else if (src1[x - 1] != '\\' && src1[x - 1] != '/') {
	    p = ap_pstrcat(a, src1, "\\", src2, NULL);
    }
    else {
	    p = ap_pstrcat(a, src1, src2, NULL);
    }

    q = p ;
    while (*q)
	{
        if (*q == '/') {
	        *q = '\\' ;
        }
	    ++q;
	}

    return p ;
#else
    return ap_make_full_path(a, src1, src2);
#endif
}

/*******************************************************************************
 * Return absolute path to file in either "regular" FCGI socket directory or
 * the dynamic directory.  Result is allocated in pool p.
 */
const char *
fcgi_util_socket_make_path_absolute(pool * const p,
        const char *const file, const int dynamic)
{
#ifdef APACHE2
    if (ap_os_is_path_absolute(p, (char *) file))
#else
    if (ap_os_is_path_absolute(file))
#endif
    {
	    return file;
    }
    else
    {
        const char * parent_dir = dynamic ? fcgi_dynamic_dir : fcgi_socket_dir;
        return (const char *) make_full_path(p, parent_dir, file);
    }
}

#ifndef WIN32
/*******************************************************************************
 * Build a Domain Socket Address structure, and calculate its size.
 * The error message is allocated from the pool p.  If you don't want the
 * struct sockaddr_un also allocated from p, pass it preallocated (!=NULL).
 */
const char *
fcgi_util_socket_make_domain_addr(pool *p, struct sockaddr_un **socket_addr,
        int *socket_addr_len, const char *socket_path)
{
    int socket_pathLen = strlen(socket_path);

    if (socket_pathLen >= sizeof((*socket_addr)->sun_path)) {
        return ap_pstrcat(p, "path \"", socket_path,
                       "\" is too long for a Domain socket", NULL);
    }

    if (*socket_addr == NULL)
        *socket_addr = ap_pcalloc(p, sizeof(struct sockaddr_un));
    else
        memset(*socket_addr, 0, sizeof(struct sockaddr_un));

    (*socket_addr)->sun_family = AF_UNIX;
    strcpy((*socket_addr)->sun_path, socket_path);

    *socket_addr_len = SUN_LEN(*socket_addr);
    return NULL;
}
#endif

/*******************************************************************************
 * Convert a hostname to an in_addr struct.
 */
int
convert_hostname_to_in_addr(const char * const hostname, struct in_addr * const addr, int *ttl)
{
    struct hostent *hp;
    int count;

    if ((hp = gethostbyname((char *)hostname)) == NULL)
        return -1;

    memcpy((char *) addr, hp->h_addr, hp->h_length);
    count = 0;
    while (hp->h_addr_list[count] != 0)
        count++;

    if (count == 0)
        return 0;
    *ttl = 60; /* Hard-coded 60 seconds to live, that's the AWS ELB value */
    return 1;
}


/*******************************************************************************
 * Convert a hostname or IP address string to an in_addr struct.
 */
static int
convert_string_to_in_addr(const char * const hostname, struct in_addr * const addr, int *ttl)
{
    struct hostent *hp;
    int count;

    addr->s_addr = inet_addr((char *)hostname);

#if !defined(INADDR_NONE) && defined(APACHE2)
#define INADDR_NONE APR_INADDR_NONE
#endif

    if (addr->s_addr == INADDR_NONE) {
        return convert_hostname_to_in_addr(hostname, addr, ttl);
    }
    return 1;
}


/*******************************************************************************
 * Build an Inet Socket Address structure, and calculate its size.
 * The error message is allocated from the pool p. If you don't want the
 * struct sockaddr_in also allocated from p, pass it preallocated (!=NULL).
 */
const char *
fcgi_util_socket_make_inet_addr(pool *p, struct sockaddr_in **socket_addr,
        int *socket_addr_len, const char *host, unsigned short port,
        int *ttl)
{
    if (*socket_addr == NULL)
        *socket_addr = ap_pcalloc(p, sizeof(struct sockaddr_in));
    else
        memset(*socket_addr, 0, sizeof(struct sockaddr_in));

    (*socket_addr)->sin_family = AF_INET;
    (*socket_addr)->sin_port = htons(port);

    /* Get an in_addr represention of the host */
    *ttl = 0;
    if (host != NULL) {
      if (convert_string_to_in_addr(host, &(*socket_addr)->sin_addr, ttl) != 1) {
            return ap_pstrcat(p, "failed to resolve \"", host,
                           "\" to exactly one IP address", NULL);
        }
    } else {
      (*socket_addr)->sin_addr.s_addr = htonl(INADDR_ANY);
    }

    *socket_addr_len = sizeof(struct sockaddr_in);
    return NULL;
}

/*******************************************************************************
 * Determine if a process with uid/gid can access a file with mode permissions.
 */
const char *
fcgi_util_check_access(pool *tp,
        const char * const path, const struct stat *statBuf,
        const int mode, const uid_t uid, const gid_t gid)
{
    struct stat myStatBuf;

    if (statBuf == NULL) {
        if (stat(path, &myStatBuf) < 0)
            return ap_psprintf(tp, "stat(%s) failed: %s", path, strerror(errno));
        statBuf = &myStatBuf;
    }

#ifndef WIN32
    /* If the uid owns the file, check the owner bits */
    if (uid == statBuf->st_uid) {
        if (mode & R_OK && !(statBuf->st_mode & S_IRUSR))
            return "read not allowed by owner";
        if (mode & W_OK && !(statBuf->st_mode & S_IWUSR))
            return "write not allowed by owner";
        if (mode & X_OK && !(statBuf->st_mode & S_IXUSR))
            return "execute not allowed by owner";
        return NULL;
    }
#else
    if (mode & _S_IREAD && !(statBuf->st_mode & _S_IREAD))
        return "read not allowed";
    if (mode & _S_IWRITE && !(statBuf->st_mode & _S_IWRITE))
        return "write not allowed";

    /* I don't think this works on FAT, but since I don't know how to check..
     * if (mode & _S_IEXEC && !(statBuf->st_mode & _S_IEXEC))
     *     return "execute not allowed"; */
#endif

#if  !defined(__EMX__) && !defined(WIN32)
    /* If the gid is same as the file's group, check the group bits */
    if (gid == statBuf->st_gid) {
        if (mode & R_OK && !(statBuf->st_mode & S_IRGRP))
            return "read not allowed by group";
        if (mode & W_OK && !(statBuf->st_mode & S_IWGRP))
            return "write not allowed by group";
        if (mode & X_OK && !(statBuf->st_mode & S_IXGRP))
            return "execute not allowed by group";
        return NULL;
    }

    /* Get the user membership for the file's group.  If the
     * uid is a member, check the group bits. */
    {
        const struct group * const gr = getgrgid(statBuf->st_gid);
        const struct passwd * const pw = getpwuid(uid);

        if (gr != NULL && pw != NULL) {
            char **user = gr->gr_mem;
            for ( ; *user != NULL; user++) {
                if (strcmp(*user, pw->pw_name) == 0) {
                    if (mode & R_OK && !(statBuf->st_mode & S_IRGRP))
                        return "read not allowed by group";
                    if (mode & W_OK && !(statBuf->st_mode & S_IWGRP))
                        return "write not allowed by group";
                    if (mode & X_OK && !(statBuf->st_mode & S_IXGRP))
                        return "execute not allowed by group";
                    return NULL;
                }
            }
        }
    }

    /* That just leaves the other bits.. */
    if (mode & R_OK && !(statBuf->st_mode & S_IROTH))
        return "read not allowed";
    if (mode & W_OK && !(statBuf->st_mode & S_IWOTH))
        return "write not allowed";
    if (mode & X_OK && !(statBuf->st_mode & S_IXOTH))
        return "execute not allowed";
#endif

    return NULL;
}


/*******************************************************************************
 * Find a FastCGI server with a matching fs_path, and if fcgi_wrapper is
 * enabled with matching uid and gid.
 */
fcgi_server *
fcgi_util_fs_get_by_id(const char *ePath, uid_t uid, gid_t gid)
{
    char path[FCGI_MAXPATH];
    fcgi_server *s;

    /* @@@ This should now be done in the loop below */
    ap_cpystrn(path, ePath, FCGI_MAXPATH);
    ap_no2slash(path);

    for (s = fcgi_servers; s != NULL; s = s->next) {
        int i;
        const char *fs_path = s->fs_path;
        for (i = 0; fs_path[i] && path[i]; ++i) {
            if (fs_path[i] != path[i]) {
                break;
            }
        }
        if (fs_path[i]) {
            continue;
        }
        if (path[i] == '\0' || path[i] == '/') {
        if (fcgi_wrapper == NULL || (uid == s->uid && gid == s->gid))
            return s;
        }
    }
    return NULL;
}

/*******************************************************************************
 * Find a FastCGI server with a matching fs_path, and if fcgi_wrapper is
 * enabled with matching user and group.
 */
fcgi_server *
fcgi_util_fs_get(const char *ePath, const char *user, const char *group)
{
    char path[FCGI_MAXPATH];
    fcgi_server *s;

    ap_cpystrn(path, ePath, FCGI_MAXPATH);
    ap_no2slash(path);

    for (s = fcgi_servers; s != NULL; s = s->next) {
        if (strcmp(s->fs_path, path) == 0) {
            if (fcgi_wrapper == NULL)
                return s;

            if (strcmp(user, s->user) == 0
                && (user[0] == '~' || strcmp(group, s->group) == 0))
            {
                return s;
            }
        }
    }
    return NULL;
}

const char *
fcgi_util_fs_is_path_ok(pool * const p, const char * const fs_path, struct stat *finfo)
{
    const char *err;

    if (finfo == NULL) {
        finfo = (struct stat *)ap_palloc(p, sizeof(struct stat));
        if (stat(fs_path, finfo) < 0)
            return ap_psprintf(p, "stat(%s) failed: %s", fs_path, strerror(errno));
    }

    if (finfo->st_mode == 0)
        return ap_psprintf(p, "script not found or unable to stat()");

    if (S_ISDIR(finfo->st_mode))
        return ap_psprintf(p, "script is a directory!");

    /* Let the wrapper determine what it can and can't execute */
    if (! fcgi_wrapper)
    {
#ifdef WIN32
        err = fcgi_util_check_access(p, fs_path, finfo, _S_IEXEC, fcgi_user_id, fcgi_group_id);
#else
        err = fcgi_util_check_access(p, fs_path, finfo, X_OK, fcgi_user_id, fcgi_group_id);
#endif
        if (err) {
            return ap_psprintf(p,
                "access for server (uid %ld, gid %ld) not allowed: %s",
                (long)fcgi_user_id, (long)fcgi_group_id, err);
        }
    }

    return NULL;
}



/*******************************************************************************
 * Allocate a new FastCGI server record from pool p with default values.
 */
fcgi_server *
fcgi_util_fs_new(pool *p)
{
    fcgi_server *s = (fcgi_server *) ap_pcalloc(p, sizeof(fcgi_server));

    /* Initialize anything who's init state is not zeroizzzzed */
    s->listenQueueDepth = FCGI_DEFAULT_LISTEN_Q;
    s->appConnectTimeout = FCGI_DEFAULT_APP_CONN_TIMEOUT;
    s->idle_timeout = FCGI_DEFAULT_IDLE_TIMEOUT;
    s->initStartDelay = DEFAULT_INIT_START_DELAY;
    s->restartDelay = FCGI_DEFAULT_RESTART_DELAY;
	s->minServerLife = FCGI_DEFAULT_MIN_SERVER_LIFE;
    s->restartOnExit = FALSE;
    s->directive = APP_CLASS_UNKNOWN;
    s->processPriority = FCGI_DEFAULT_PRIORITY;
    s->envp = &fcgi_empty_env;

#ifdef WIN32
    s->listenFd = (int) INVALID_HANDLE_VALUE;
#else
    s->listenFd = -2;
#endif

    return s;
}

/*******************************************************************************
 * Add the server to the linked list of FastCGI servers.
 */
void
fcgi_util_fs_add(fcgi_server *s)
{
    s->next = fcgi_servers;
    fcgi_servers = s;
}

/*******************************************************************************
 * Configure uid, gid, user, group, username for wrapper.
 */
const char *
fcgi_util_fs_set_uid_n_gid(pool *p, fcgi_server *s, uid_t uid, gid_t gid)
{
#ifndef WIN32

    struct passwd *pw;
    struct group  *gr;

    if (fcgi_wrapper == NULL)
        return NULL;

    if (uid == 0 || gid == 0) {
        return "invalid uid or gid, see the -user and -group options";
    }

    s->uid = uid;
    pw = getpwuid(uid);
    if (pw == NULL) {
        return ap_psprintf(p,
            "getpwuid() couldn't determine the username for uid '%ld', "
            "you probably need to modify the User directive: %s",
            (long)uid, strerror(errno));
    }
    s->user = ap_pstrdup(p, pw->pw_name);
    s->username = s->user;

    s->gid = gid;
    gr = getgrgid(gid);
    if (gr == NULL) {
        return ap_psprintf(p,
            "getgrgid() couldn't determine the group name for gid '%ld', "
            "you probably need to modify the Group directive: %s",
            (long)gid, strerror(errno));
    }
    s->group = ap_pstrdup(p, gr->gr_name);

#endif /* !WIN32 */

    return NULL;
}

/*******************************************************************************
 * Allocate an array of ServerProcess records.
 */
ServerProcess *
fcgi_util_fs_create_procs(pool *p, int num)
{
    int i;
    ServerProcess *proc = (ServerProcess *)ap_pcalloc(p, sizeof(ServerProcess) * num);

    for (i = 0; i < num; i++) {
#ifdef WIN32
        proc[i].handle = INVALID_HANDLE_VALUE;
        proc[i].terminationEvent = INVALID_HANDLE_VALUE;
#endif
        proc[i].pid = 0;
        proc[i].state = FCGI_READY_STATE;
    }
    return proc;
}

int fcgi_util_ticks(struct timeval * tv)
{
#ifdef WIN32
    /* millisecs is sufficent granularity */
    DWORD millis = GetTickCount();

    tv->tv_sec = millis / 1000;
    tv->tv_usec = (millis % 1000) * 1000;

    return 0;
#else
    return gettimeofday(tv, NULL);
#endif
}
